Partnership model investigation

ethax-ross opened the pull request:
[Jira-4215](https://dfedigital.atlassian.net.mcas.ms/browse/CPDLP-4215)

- Add new partnership models

Adds the new `LeadProviderActivePeriod` and `LeadProviderDeliveryPartnership` models along with minimal validation rules.

- Add school and lead_provider_delivery_partnership to SchoolPartnership

These two relationships define the new way to retrieve the lead provider, delivery partner and registration period for a `SchoolPartnership`. The direct relationships this replaces will be removed in a following commit.

- Replace direct relationships on SchoolPartnership

Remove the direct relationships to `lead_provider`, `delivery_partner` and `registration_period`. Going forward, these will be via `lead_provider_delivery_partnership` and `lead_provider_active_period` to allow greater flexibility in how schools engage with lead providers and delivery partners across registration periods.

- Add expression_of_interest to TrainingPeriod

Add the `expression_of_interest` to the `TrainingPeriod` model. Going forward the `TrainingPeriod` will require an `expression_of_interest` or `school_partnership` (or both, but at least one).

## Notes

- `LeadProviderActivePeriod` is used to define which of the lead providers are active for a given registration year.
- `LeadProviderDeliveryPartnership` describes which delivery partners are associated with which lead provider for each registration year.

These two models will be created up-front by RECT to align with how we setup a new academic year in ECF at the moment (which I believe involves importing the active lead providers and delivery partners from a CSV).

- `TrainingPeriod` describes a period in which a participant is in training. 
- Think of this as the induction record replacement to show how a participant's training changes over time.
- It will have an `expression_of_interest` or a `school_partnership` (or both, but at least one).
- When a participant is on-boarded their training period will either be:
  - Assigned to a `SchoolPartnership` through `school_partnership` if one already exists for the desired school/delivery partner.
  - Assigned to a `LeadProviderActivePeriod` as an `expression_of_interest`
 - Once the lead provider confirms an `expression_of_interest` a `SchoolPartnership` will be created and the `school_partnership` assigned.

We are currently expecting one delivery partner per school per lead provider, but in the future/after looking at the existing data we may need to support multiple. The schema would allow for that.

```mermaid
erDiagram
    LeadProvider ||--|{ ActiveLeadProvider : ""
    RegistrationPeriod ||--|{ ActiveLeadProvider : ""
    ActiveLeadProvider ||--|{ LeadProviderDeliveryPartnership : ""
    DeliveryPartner ||--|{ LeadProviderDeliveryPartnership : ""
    LeadProviderDeliveryPartnership ||--|{ SchoolPartnership : ""
    ActiveLeadProvider |o..|{ TrainingPeriod : "expression_of_interest"
    School ||--|{ SchoolPartnership : ""
    SchoolPartnership |o..|{ TrainingPeriod : "confirmed_school_partnership"
```

ethax-ross committed:
Add new partnership models

Adds the new `LeadProviderActivePeriod` and `LeadProviderDeliveryPartnership`
models along with minimal validation rules.

ethax-ross committed:
Add new partnership models

Adds the new `LeadProviderActivePeriod` and `LeadProviderDeliveryPartnership`
models along with minimal validation rules.

ethax-ross committed:
Add `school` and `lead_provider_delivery_partnership` to `SchoolPartnership`

These two relationships define the new way to retrieve the lead provider,
delivery partner and registration period for a `SchoolPartnership`. The direct
relationships this replaces will be removed in a following commit.

ethax-ross committed:
Replace direct relationships on `SchoolPartnership`

Remove the direct relationships to `lead_provider`, `delivery_partner` and
`registration_period`. Going forward, these will be via
`lead_provider_delivery_partnership` and `lead_provider_active_period` to
allow greater flexibility in how schools engage with lead providers and
delivery partners across registration periods.

ethax-ross committed:
Add `expression_of_interest` to `TrainingPeriod`

Add the `expression_of_interest` to the `TrainingPeriod` model. Going
forward the `TrainingPeriod` will require an `expression_of_interest` or
`school_partnership` (or both, but at least one).

ethax-ross committed:
Add uniqueness constraints to partnership models

To ensure only single combinations of the various partnership models can be
created.

cwrw reviewed (commented):
thanks Ross looks great!

leandroalemao reviewed (approved):
Well done @ethax-ross tks

ethax-ross committed:
Ensure `SchoolPartnership` school is consistent with school periods

We link a `TrainingPeriod` to a school already via one of the school periods.
When connecting a `TrainingPeriod` to a `SchoolPartnership` we need to ensure
the `school` of the partnership is consistent with the school periods.

We need to duplicate this relationship with a `School` due to the timings of
how partnerships and training periods can be created (either one could come
first and needs a reference to the school, and they can both exist
independently).

leandroalemao committed:
Add gems to support the API

leandroalemao committed:
Add APIToken

leandroalemao committed:
Add partnerships endpoints

ethax-ross committed:
Support creating/updating partnerships

Add support for creating and updating partnerships by replicating the service behaviour in ECF.

Add `ecf_id` to `School`, `DeliveryPartner` and `LeadProvider` models.

Add convenience methods to `School` for eligibility checks.

Add `Create` and `Update` services.

Accept expressions of interest for the same school when creating a `SchoolPartnership`.

leandroalemao committed:
Remove pagy from Gemfile as it is auto pulled in by govuk-components

leandroalemao committed:
Add token authentication to requests specs

ethax-ross committed:
Ignore expressions of interest that have been actioned

We only want to create partnerships for the expressions of interest that have
not yet been accepted (have no `school_partnership`).

ethax-ross committed:
Merge pull request #549 from DFE-Digital/partnerships-create-update

Support creating/updating partnerships

leandroalemao committed:
Remove rebase issues

leandroalemao committed:
Merge pull request #544 from DFE-Digital/partnership-endpoints-investigation

Partnership endpoints investigation

ethax-ross commented:
Closing spike as we've moved on to statements
